Eclipse resource filters

kpage opened the pull request:
Hi @oehme ! This is a draft pull request for the first phase of the eclipse resource filters, adding the EclipseProject DSL support for filters with a basic custom matcher.  I am still cleaning up the documentation, but I wanted to open this so you could start to take a look.

Any of the checked boxes below indicate that I took action:

- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [x] Before submitting a pull request, I started a discussion on the [Gradle eclipse Buildship plugin bug tracker](https://bugs.eclipse.org/bugs/show_bug.cgi?id=495799)
- [x] Design document: https://github.com/gradle/gradle/tree/master/design-docs/features/ide-integration/richer-eclipse-model/resource-filters
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on your local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [x] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

kpage commented:
Some notes:

* Deviation from spec: I used random UUIDs for the resource filters instead of epoch time, as the filters were all created at about the same time and it looked like collisions were likely.
* I tried a few methods of testing the generated filter XML.  In the end, as the XML can get pretty complex, it seemed like the easiest way was to just assert a block of XML directly in the EclipseProjectIntegrationTest.  Because the filter ids are autogenerated, they are stripped out before asserting if they were not passed in the expected XML.
* Still need to implement reading filters from an existing .project file

kpage commented on subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseProjectFixture.groovy:
```diff
@@ -40,10 +46,24 @@ class EclipseProjectFixture {
         assert this.project.projects.project*.text() == referencedProjects as List
     }
 
+    void assertHasResourceFilterXml(String expectedFilteredResourcesXml) {
```
Now that the ids are sequential and predictable, most of this code is now unnecessary as we can just put the expected id in the xml.  I will remove it.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseProject.java:
```diff
@@ -124,6 +124,8 @@
 
     private Set<Link> linkedResources = Sets.newLinkedHashSet();
 
+    private ResourceFilterSet resourceFilterSet = new ResourceFilterSet();
```
Ok, sounds good.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Project.java:
```diff
@@ -242,6 +259,63 @@ private void addLinkedResourcesToXml() {
     }
 
+    private void addResourceFiltersToXml() {
+        // other nodes like linkedResources are appended even if they are empty, but as
```
Ok, I will update the old failing tests that were not expecting this empty node.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Project.java:
```diff
@@ -242,6 +259,63 @@ private void addLinkedResourcesToXml() {
+
+    private int getResourceFilterType(ResourceFilter resourceFilter) {
+        int type = 0;
+        if (resourceFilter == null) {
```
Ok, I will pick and document some default values and prevent nulls in the ResourceFilter setters.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,147 @@
+ *
+ * @since NEXT_VERSION
+ */
+public class ResourceFilter extends GroovyObjectSupport {
```
Ok

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,147 @@
+public class ResourceFilter extends GroovyObjectSupport {
+    // Originally the spec mentioned using current epoch time, but all the filters tend to be created
+    // in the same closure and collisions were possible, so a random UUID was used instead.
+    private String id = UUID.randomUUID().toString();
```
Will remove as per your other comment about using sequential ids.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,147 @@
+    private String id = UUID.randomUUID().toString();
+    private ResourceFilterAppliesTo appliesTo;
+    private ResourceFilterType type;
+    private Boolean recursive;
```
Ok

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,147 @@
+    }
+    
+    public void setAppliesTo(ResourceFilterAppliesTo appliesTo) {
+        this.appliesTo = appliesTo;
```
Ok, would InvalidUserDataException be appropriate here?

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,147 @@
+        this.matcher = matcher;
+    }
+
+    public void matcher(Closure configureClosure) {
```
Ok, I see some other examples of using both an Action and a ClosureBackedAction in the code, I'll do it like that.  Should I also add an Action overload for EclipseProject.resourceFilter?

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterAppliesTo.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @since NEXT_VERSION
+ */
+public enum ResourceFilterAppliesTo {
```
Would you like to sync the naming with the DSL?  I used ResourceFilterAppliesTo as "appliesTo" was used in the spec DSL.  Would you rather use "ResourceFilterTarget" and "target"?

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @since NEXT_VERSION
+ */
+public class ResourceFilterMatcher extends GroovyObjectSupport {
```
Ok.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Project.java:
```diff
@@ -242,6 +259,63 @@ private void addLinkedResourcesToXml() {
+                    break;
+            }
+        }
+        if (resourceFilter.getAppliesTo() != null) {
```
ok

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Project.java:
```diff
@@ -242,6 +259,63 @@ private void addLinkedResourcesToXml() {
+        if (resourceFilter == null) {
+            return type;
+        }
+        if (resourceFilter.getType() != null) {
```
ok

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/Project.java:
```diff
@@ -242,6 +259,63 @@ private void addLinkedResourcesToXml() {
+                    break;
+            }
+        }
+        if (resourceFilter.getRecursive() != null && resourceFilter.getRecursive()) {
```
ok

kpage commented:
Hi @oehme , I have addressed your comments with a new commit.  We can squash later before merging.  I will work on the reading of the existing .project filteredResources next and add as a new commit here.

kpage commented:
Hi @oehme , I have now added the support for reading filteredResources from an existing .project file and documentation for the new features.  Would you please take another look?

- Let me know if you want to do anything about the target/appliesTo (files/folders) naming.  I made them all consistent (“appliesTo”) for now, but we can change it if you wish.
- I wasn’t sure how the gradle team manages @since tags.  For now I used the placeholder NEXT_VERSION.  Please let me know if I need to update these.
- I added extra commits to make the changes easier to review, but we can squash and rebase before merging.

oehme commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * The model of an Eclipse resource filter matcher.
+ * <p>
+ * The matcher configures the behavior of its parent ResourceFilter.  The matcher configures
```
The matcher does not configure the behavior. The matcher decides when the containing filter (or containing matcher) applies.

kpage replied:
Updated

oehme commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+    private String arguments;
+    private Set<ResourceFilterMatcher> children = Sets.newLinkedHashSet();
+
+    // Needed for DSL configuration Actions
```
Unnecessary comment

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,195 @@
+    private boolean recursive = true;
+    private ResourceFilterMatcher matcher;
+
+    // Needed for DSL configuration Actions
```
Ok, will remove the comment.  I was trying to find a way to make sure the class was in a valid state by validating the setters and requiring a constructor with all arguments, but this doesn't seem to work well with configuration by closure/Action.

oehme replied:
Yes, we can't validate this up-front, we can only give reasonable defaults (or fail later on if the user forgot something)

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,195 @@
+/**
+ * The gradle DSL model of an Eclipse resource filter.
+ *
+ * Later this will be extended with additional helpers, but for now it just allows specifying a filter with a custom matcher
```
Ok, removed :)

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,195 @@
+ *
+ * eclipse {
+ *   project {
+ *     // the following closure configures the ResourceFilter
```
Removed comment

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,195 @@
+    }
+    
+    /**
+     * Sets the ResourceFilterAppliesTo
```
Ok, copied the getter documentation to the setter also

kpage commented:
Ok, I have added test "existing project file with equivalent resource filters is unchanged" to test that an existing .project file with two resource filters is unchanged after the "eclipse" task is run on a gradle project that defines two equivalent filters.

One thing to note is that the ids of the filters will possibly change if Eclipse had generated the filters, as it will use a timestamp instead of sequential integers.  As we do not store the id in our model, we can't recreate it, so the id will just change to a sequential number.  But subsequent runs of the "eclipse" task will keep the same filter ordering and numbering.  Do you think this will be an issue?

kpage commented:
Hi @oehme , I have completed the latest round of code review changes, would you please have another look?  Thanks!

oehme commented:
> One thing to note is that the ids of the filters will possibly change if Eclipse had generated the filters, as it will use a timestamp instead of sequential integers. 

I think we can live with that, yes. Can you check that Eclipse doesn't change the id back when you import the project? Otherwise it could become an annoying back-and-forth.

Apart from that, this looks really great to me and I can merge it as soon as the question above is clarified.

kpage commented:
Hi @oehme , I tested importing the generated .project file into Eclipse, and Eclipse does not change the identifiers.  It will only change the ids if the filters are modified in the UI, and in that case it rewrites all of them (sometimes the ids are actually not unique, they are the same exact timestamp...).  So I think this behavior is fine.  

I have squashed the code review commits and rebased onto the latest master, you can merge when you are ready.

oehme commented:
Thanks for the great work @kpage! I'll be out for a week, but will merge this when I'm back. It will be part of Gradle 3.4, so if you have some time to adjust the "NEXT_VERSION" tags that would help.

kpage commented:
Hi @oehme , I changed the NEXT_VERSIONs to 3.4 and rebased onto latest master.

eriwen commented:
Hi @kpage. This change looks good but we do have an outstanding test failure on Windows `EclipseProjectIntegrationTest.existing project file with equivalent resource filters is unchanged`. Would you please look into fixing it? 

<details><summary>Failure details</summary>
Condition not satisfied:

projectFileOriginalText == projectFile.text
|                       |  |           |
|                       |  |           <?xml version="1.0" encoding="UTF-8"?>
|                       |  |           <projectDescription>
|                       |  |               <name>root</name>
|                       |  |               <comment></comment>
|                       |  |               <projects/>
|                       |  |               <natures>
|                       |  |                   <nature>org.eclipse.jdt.core.javanature</nature>
|                       |  |               </natures>
|                       |  |               <buildSpec>
|                       |  |                   <buildCommand>
|                       |  |                       <name>org.eclipse.jdt.core.javabuilder</name>
|                       |  |                       <arguments/>
|                       |  |                   </buildCommand>
|                       |  |               </buildSpec>
|                       |  |               <linkedResources/>
|                       |  |               <filteredResources>
|                       |  |                   <filter>
|                       |  |                       <id>1</id>
|                       |  |                       <type>9</type>
|                       |  |                       <name/>
|                       |  |                       <matcher>
|                       |  |                           <id>org.eclipse.some.custom.matcher</id>
|                       |  |                           <arguments>foobar</arguments>
|                       |  |                       </matcher>
|                       |  |                   </filter>
|                       |  |                   <filter>
|                       |  |                       <id>2</id>
|                       |  |                       <type>30</type>
|                       |  |                       <name/>
|                       |  |                       <matcher>
|                       |  |                           <id>org.eclipse.some.custom.matcher</id>
|                       |  |                           <arguments>barfoo</arguments>
|                       |  |                       </matcher>
|                       |  |                   </filter>
|                       |  |               </filteredResources>
|                       |  |           </projectDescription>
|                       |  C:\tcagent1\work\4b92f910977a653d\subprojects\ide\build\tmp\test files\EclipseProjectIntegrationTest\existing_project_fi...unchanged\j33k8\.project
|                       false
|                       36 differences (95% similarity)
|                       <?xml version="1.0" encoding="UTF-8"?>(-~)\n<projectDescription>(-~)\n\t<name>root</name>(-~)\n\t<comment></comment>(-~)\n\t<projects/>(-~)\n\t<natures>(-~)\n\t\t<nature>org.eclipse.jdt.core.javanature</nature>(-~)\n\t</natures>(-~)\n\t<buildSpec>(-~)\n\t\t<buildCommand>(-~)\n\t\t\t<name>org.eclipse.jdt.core.javabuilder</name>(-~)\n\t\t\t<arguments/>(-~)\n\t\t</buildCommand>(-~)\n\t</buildSpec>(-~)\n\t<linkedResources/>(-~)\n\t<filteredResources>(-~)\n\t\t<filter>(-~)\n\t\t\t<id>1</id>(-~)\n\t\t\t<type>9</type>(-~)\n\t\t\t<name/>(-~)\n\t\t\t<matcher>(-~)\n\t\t\t\t<id>org.eclipse.some.custom.matcher</id>(-~)\n\t\t\t\t<arguments>foobar</arguments>(-~)\n\t\t\t</matcher>(-~)\n\t\t</filter>(-~)\n\t\t<filter>(-~)\n\t\t\t<id>2</id>(-~)\n\t\t\t<type>30</type>(-~)\n\t\t\t<name/>(-~)\n\t\t\t<matcher>(-~)\n\t\t\t\t<id>org.eclipse.some.custom.matcher</id>(-~)\n\t\t\t\t<arguments>barfoo</arguments>(-~)\n\t\t\t</matcher>(-~)\n\t\t</filter>(-~)\n\t</filteredResources>(-~)\n</projectDescription>(-~)\n
|                       <?xml version="1.0" encoding="UTF-8"?>(\r)\n<projectDescription>(\r)\n\t<name>root</name>(\r)\n\t<comment></comment>(\r)\n\t<projects/>(\r)\n\t<natures>(\r)\n\t\t<nature>org.eclipse.jdt.core.javanature</nature>(\r)\n\t</natures>(\r)\n\t<buildSpec>(\r)\n\t\t<buildCommand>(\r)\n\t\t\t<name>org.eclipse.jdt.core.javabuilder</name>(\r)\n\t\t\t<arguments/>(\r)\n\t\t</buildCommand>(\r)\n\t</buildSpec>(\r)\n\t<linkedResources/>(\r)\n\t<filteredResources>(\r)\n\t\t<filter>(\r)\n\t\t\t<id>1</id>(\r)\n\t\t\t<type>9</type>(\r)\n\t\t\t<name/>(\r)\n\t\t\t<matcher>(\r)\n\t\t\t\t<id>org.eclipse.some.custom.matcher</id>(\r)\n\t\t\t\t<arguments>foobar</arguments>(\r)\n\t\t\t</matcher>(\r)\n\t\t</filter>(\r)\n\t\t<filter>(\r)\n\t\t\t<id>2</id>(\r)\n\t\t\t<type>30</type>(\r)\n\t\t\t<name/>(\r)\n\t\t\t<matcher>(\r)\n\t\t\t\t<id>org.eclipse.some.custom.matcher</id>(\r)\n\t\t\t\t<arguments>barfoo</arguments>(\r)\n\t\t\t</matcher>(\r)\n\t\t</filter>(\r)\n\t</filteredResources>(\r)\n</projectDescription>(\r)\n
<?xml version="1.0" encoding="UTF-8"?>
<projectDescription>
    <name>root</name>
    <comment></comment>
    <projects/>
    <natures>
        <nature>org.eclipse.jdt.core.javanature</nature>
    </natures>
    <buildSpec>
        <buildCommand>
            <name>org.eclipse.jdt.core.javabuilder</name>
            <arguments/>
        </buildCommand>
    </buildSpec>
    <linkedResources/>
    <filteredResources>
        <filter>
            <id>1</id>
            <type>9</type>
            <name/>
            <matcher>
                <id>org.eclipse.some.custom.matcher</id>
                <arguments>foobar</arguments>
            </matcher>
        </filter>
        <filter>
            <id>2</id>
            <type>30</type>
            <name/>
            <matcher>
                <id>org.eclipse.some.custom.matcher</id>
                <arguments>barfoo</arguments>
            </matcher>
        </filter>
    </filteredResources>
</projectDescription>

    at org.gradle.plugins.ide.eclipse.EclipseProjectIntegrationTest.existing project file with equivalent resource filters is unchanged(EclipseProjectIntegrationTest.groovy:555)

------- Stdout: -------
:eclipseClasspath

:eclipseJdt

:eclipseProject

:eclipse


BUILD SUCCESSFUL


Total time: 0.089 secs

------- Stderr: -------
Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://javax.xml.XMLConstants/property/accessExternalDTD' is not recognized.

Warning:  org.apache.xerces.parsers.SAXParser: Property 'http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit' is not recognized.
</details>

oehme commented on subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseProjectIntegrationTest.groovy:
```diff
@@ -78,6 +78,540 @@ eclipse {
+
+        '''
+        project.assertHasResourceFilterXml(resourceFilterXml)
+        projectFileOriginalText == projectFile.text
```
This is the reason for the test failure. You need to normalize line separators if you want to compare a String and the content from a file.

kpage commented:
Thanks @eriwen  and @oehme , I have fixed the test and rebased onto the latest master.

Does gradle org have a public CI for pull requests?  I ran the tests locally before on Linux but did not have a Windows machine available.

eriwen commented:
@kpage Yeah, I kicked off CI for your PR here: https://builds.gradle.org/viewQueued.html?itemId=2197344&tab=queuedBuildOverviewTab

I'm afraid this didn't make it into 3.4. Could you please update the `@since` annotations to 3.5 in the meantime?

kpage commented:
Hi @eriwen , I have updated the new `@since` annotations to 3.5

eriwen commented:
@kpage The CI build passed and with @oehme's +1 I'm going to merge this. Thanks very much for your contributions!

oehme commented:
@kpage Would you like to follow up with the Tooling API part next so we have a complete story for 3.5?

big-guy commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+        return children;
+    }
+
+    public void setChildren(Set<ResourceFilterMatcher> children) {
```
Is there an example of when this would be useful?

One weird thing about the way this is done is `matcher {}` at this level _adds_ a new matcher, but `matcher {}` in the resource filter configures an existing one.

kpage replied:
Hi @big-guy , are you asking why nested `ResourceFilterMatcher`s are supported?

This is required by the existing Eclipse model of filters/matchers.  Nested matchers are actually a pretty common use case.  Suppose you wanted to filter out either of two folders from showing in Eclipse, this would be modeled as an `or` matcher with nested matchers for the folder names:

```
eclipse {
  project {
    resourceFilter {
      matcher {
        id = 'org.eclipse.ui.ide.orFilterMatcher'
        matcher {
          id = 'org.eclipse.ui.ide.multiFilter'
          arguments = '1.0-name-matches-false-false-node_modules'
        }
        matcher {
          id = 'org.eclipse.ui.ide.multiFilter'
          arguments = '1.0-name-matches-false-false-build'
        }
      }
    }
  }
}
```

Regarding the DSL issue you mention where `matcher` at the filter level creates/configures a single matcher whereas `matcher` at the matcher level creates a new nested matcher, that is something I struggled with as well.  

I took a look at other examples in the gradle code where the DSL configures single objects vs. collections of objects, and this seemed to be how it was done elsewhere (`EclipseProject.linkedResource`, for example, creates a new Link and adds to the set whereas `EclipseClasspath.file` configures the single `XmlFileContentMerger`).  

The confusion here just seems to stem from the required domain complexity that a filter can have exactly one matcher but matchers themselves can have multiple nested matchers.  I guess the method names could be differentiated (`matcher` vs `nestedMatcher`?), but that might also cause another confusion as to why the same object type is constructed with different keywords depending on its placement in the DSL tree.  Let me know if you have any suggestions.

Please keep in mind that the current implementation is just the "raw" domain model that gets gradle to parity with the Eclipse resource filters, but @oehme [has spec'd a much more intuitive extension to the DSL](https://github.com/gradle/gradle/tree/master/design-docs/features/ide-integration/richer-eclipse-model/resource-filters) that can be layered on top of this.

big-guy commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+    private String arguments;
+    private Set<ResourceFilterMatcher> children = Sets.newLinkedHashSet();
+
+    public ResourceFilterMatcher() {
```
Several of these new classes have 2 constructors, but it seems like only one of them is useful/used often?

kpage replied:
The default constructor was required to initialize an empty object that is configured by the Action while the full constructor is useful to construct instances for tests (ProjectTest, ResourceFilterTest, ResourceFilterMatcherTest).  Please let me know if you have a more idiomatic way of constructing the DSL objects for Actions, etc.

big-guy commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+        this.children = children;
+    }
+
+    public String getId() {
```
This isn't marked as `@Nullable`, but it seems like we expect it could be null sometimes (see hashCode)?

kpage replied:
The id is required by Eclipse for the matcher to work, I could add validation logic in setId to throw InvalidUserDataException if null is passed like we do for "children".  I added the null check in hashCode because these are mutable DSL objects that are constructed empty and then populated later by Action or from reading an existing .project file, so it is possible that hashCode might be called and throw NPE before the object is fully/correctly configured.  EqualsVerifier pointed out the NPE when I was testing.

kpage replied:
Null check on setter added in #1328

big-guy commented on subprojects/ide/ide.gradle:
```diff
@@ -30,6 +30,7 @@ dependencies {
     compile libraries.inject
 
     testCompile libraries.xmlunit
+    testCompile libraries.equalsVerifier
```
This is a nit-picky thing, but we only use this in two tests and I think we could capture about the same thing by checking that the effect of a bad equals/hashCode method is caught in the integration tests (for the generated Eclipse metadata files).

kpage replied:
Ok, I had added this for myself to catch errors with the manually written equals/hashCode required by the DSL and decided to leave it in case the gradle team found it useful.

I've found EqualsVerifier quite useful on other projects as a low-effort way to catch issues with Java hashCode/equals (it catches things like possible NPEs, fields that were recently added without updating hashCode/equals, using a field in hashCode but forgetting it in equals, the semantics of getClass() vs instanceof in the presence of inheritance, and a few other useful things).  To check these edge cases in the integration tests might run over 1000 more lines of test code and take longer to execute.

If you think you will not use it often, I can remove it.  Please let me know what you decide.

kpage commented:
Hi @big-guy , as this branch has already merged, I have opened a new pull request to address changes from your comments: #1328

@oehme sure, after the new code review comments are finalized, I can look at the Tooling API.

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/EclipseProject.java:
```diff
@@ -285,6 +300,47 @@ public void linkedResource(Map<String, String> args) {
+     *
+     * @param configureClosure The closure to use to configure the resource filter.
+     */
+    public ResourceFilter resourceFilter(@DelegatesTo(value=ResourceFilter.class, strategy = Closure.DELEGATE_FIRST) Closure configureClosure) {
```
Done in #1328

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilter.java:
```diff
@@ -0,0 +1,196 @@
+     *
+     * @param configureClosure The closure to use to configure the matcher.
+     */
+    public ResourceFilterMatcher matcher(@DelegatesTo(value=ResourceFilterMatcher.class, strategy = Closure.DELEGATE_FIRST) Closure configureClosure) {
```
Done in #1328

kpage commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @param configureClosure The closure to use to configure the matcher.
+     */
+    public ResourceFilterMatcher matcher(@DelegatesTo(value=ResourceFilterMatcher.class, strategy = Closure.DELEGATE_FIRST) Closure configureClosure) {
```
Done in #1328
